src/dune_rules/merlin_server.ml

```diff
@@ -78,48 +42,70 @@ let _to_local abs_file_path =
   in
   match path_opt with
   | Some path -> (
-    try
-      Ok
-        Path.(
-          Filename.concat "." path |> of_string |> drop_optional_build_context
-          |> local_part)
+    try Ok Path.(Filename.concat "." path |> of_string |> local_part)
     with _ -> Printf.sprintf "Could not resolve path %s" path |> error )
   | None ->
     Printf.sprintf "Path is not in dune workspace %s" abs_file_path |> error
 
-let load_merlin_file dir =
+let get_merlin_files_paths local_path =
   let workspace = Workspace.workspace () in
   let context =
     Option.value ~default:Context_name.default workspace.merlin_context
   in
   let ctx = Context_name.to_string context in
-  let dir_path = Path.Build.(append_local (relative root ctx) dir) in
-  let file_path = Path.Build.relative dir_path Merlin.merlin_file_name in
-  if Path.exists (Path.build file_path) then
-    let build = Build.lines_of (Path.build file_path) in
-    let lines, _ = Build.exec build in
-    Dot_merlin.parse_lines lines
-  else
-    []
+  let ctx_root = Path.Build.(relative root ctx) in
+  let dir_path = Path.Build.(append_local ctx_root local_path) in
+  let merlin_path = Path.Build.relative dir_path Merlin.merlin_folder_name in
+  let files =
+    Result.value ~default:[] (Path.readdir_unsorted (Path.build merlin_path))
+    |> List.fast_sort ~cmp:Stdlib.compare
+  in
+  List.map files ~f:(fun f -> Path.Build.relative merlin_path f |> Path.build)
 
-let out s =
-  Dot_merlin.to_channel ~out_channel:stdout s;
-  flush stdout
+let load_merlin_file local_path file =
+  let no_config_error () =
+    Merlin_conf.make_error "Project isn't built. (Try calling `dune build`.)"
+  in
+  (* We search for an appropriate merlin configuration in the current directory
+     and its parents *)
+  let rec find_closest path =
+    let filename = String.lowercase_ascii file in
+    let file_paths = get_merlin_files_paths path in
+    let result =
+      List.find_map file_paths ~f:(fun file_path ->
+          if Path.exists file_path then
+            match Merlin.Processed.load_file file_path with
+            | Some config -> Merlin.Processed.get config ~filename
+            | None -> None
+          else
+            None)
+    in
+    match result with
+    | Some p -> Some p
+    | None ->
+      Option.bind
+        ( if Path.Local.is_root path then
+          None
+        else
+          Path.Local.parent path )
+        ~f:find_closest
+  in
+  Option.value (find_closest local_path) ~default:(no_config_error ())
```

Comment:
You might as well make `no_config_error` a value if you're going to call it here regardless.